File: lecture03/exercise03-B/solution.cpp
// Exercise 03-B: Largest Number

#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

// Functions

bool read_numbers(vector<string> &numbers) {
    numbers.clear();

    string line;
    if (getline(cin, line)) {
    	stringstream ss(line);
    	string n;
    	while (ss >> n) {
    	    numbers.push_back(n);
	}
    }

    return numbers.size();
}

// Main Execution

int main(int argc, char *argv[]) {
    vector<string> numbers;

    while (read_numbers(numbers)) {
    	sort(numbers.begin(), numbers.end(), [](const string &a, const string &b) {
    	    return stoi(a + b) > stoi(b + a);
	});

	for (auto n : numbers) {
	    cout << n;
	}
	cout << endl;
    }

    return 0;
}

File: lecture03/exercise03-B/solution.py
#!/usr/bin/env python3

# Exercise 03-B: Largest Number

import functools
import sys

# Functions

def compare_numbers(a: str, b: str) -> int:
    ab = int(a + b)
    ba = int(b + a)
    return ab - ba

# Main execution

def main() -> None:
    sort_key = functools.cmp_to_key(compare_numbers)
    for line in sys.stdin:
        print(''.join(sorted(line.split(), key=sort_key, reverse=True)))

if __name__ == '__main__':
    main()
